cpu/samd21/cpu.c

```diff
@@ -31,12 +40,27 @@ static void clk_init(void)
     PM->APBAMASK.reg = (PM_APBAMASK_PM | PM_APBAMASK_SYSCTRL |
                         PM_APBAMASK_GCLK);
 
-    /* adjust NVM wait states, see table 42.30 (p. 1070) in the datasheet */
-#if (CLOCK_CORECLOCK > 24000000)
+    /* adjust NVM wait states, see SAMD21 datasheet
+       Rev A (2017) table 37-40 , page 816 */
     PM->APBBMASK.reg |= PM_APBBMASK_NVMCTRL;
+#if (VDD_MILLIVOLTS > 2700)
+#if (CLOCK_CORECLOCK > CLOCK_24MHZ)
     NVMCTRL->CTRLB.reg |= NVMCTRL_CTRLB_RWS(1);
-    PM->APBBMASK.reg &= ~PM_APBBMASK_NVMCTRL;
+#else
+    NVMCTRL->CTRLB.reg |= NVMCTRL_CTRLB_RWS(0);
 #endif
+#else /* VDD_MILLIVOLTS <= 2700 */
+#if (CLOCK_CORECLOCK > CLOCK_42MHZ)
+    NVMCTRL->CTRLB.reg |= NVMCTRL_CTRLB_RWS(3);
+#elif (CLOCK_CORECLOCK > CLOCK_28MHZ)
+    NVMCTRL->CTRLB.reg |= NVMCTRL_CTRLB_RWS(2);
+#elif (CLOCK_CORECLOCK > CLOCK_14MHZ)
+    NVMCTRL->CTRLB.reg |= NVMCTRL_CTRLB_RWS(1);
+#else
+    NVMCTRL->CTRLB.reg |= NVMCTRL_CTRLB_RWS(0);
+#endif
```

Comment:
> I do think there should be a comment in there somehow linking the original table from the data sheet to the observed linear correlation

I couldn't agree more, these kind of lines should be well documented! I might have been sloppy here in my example code above...